File: 3260-find-the-largest-palindrome-divisible-by-k/3260-find-the-largest-palindrome-divisible-by-k.cpp
class Solution {
    int solve(int index,int currSum,int n,int k,vector<vector<int>>&dp,vector<int>&mul)
    {
        if(index>n-1-index)
            return (currSum==0);
        if(dp[index][currSum]!=-1)
            return dp[index][currSum];
        dp[index][currSum]=0;
        for(int j=(index?0:1);j<=9;j++)
        {
            int ncurrSum=currSum;
            ncurrSum=(ncurrSum+ j*mul[n-1-index] )%k;
            if(n-1-index != index)
                ncurrSum=(ncurrSum+ j*mul[index] )%k;
            dp[index][currSum]=std::max(dp[index][currSum],solve(index+1,ncurrSum,n,k,dp,mul));
        }
        return dp[index][currSum];
    }
    void solve2(int index,int currSum,int n,int k,vector<vector<int>>&dp,string& ans,vector<int>&mul)
    {
        if(index>n-1-index)
            return;
        
        for(int j=9;;j--)
        {
            int ncurrSum=currSum;
            ncurrSum=(ncurrSum+ j*mul[n-1-index] )%k;
            if(n-1-index != index)
                ncurrSum=(ncurrSum+ j*mul[index] )%k;
            if(solve(index+1,ncurrSum,n,k,dp,mul))
            {
                ans+=(char)('0'+j);
                solve2(index+1,ncurrSum,n,k,dp,ans,mul);
                return;
            }
        }
        
    }
public:
    string largestPalindrome(int n, int k) {
        vector<int> mul;
        int factor=1;
        for(int i=0;i<=n;i++)
        {
            mul.push_back(factor);
            factor=(factor*10)%k;
        }
        
        vector<vector<int>> dp(n,vector<int>(k,-1));
        solve(0,0,n,k,dp,mul);
        string ans="";
        solve2(0,0,n,k,dp,ans,mul);
        
        for(int j=n/2-1;j>=0;j--)
            ans+=ans[j];
        
        return ans;
    }
};

File: 3260-find-the-largest-palindrome-divisible-by-k/3260-find-the-largest-palindrome-divisible-by-k.java
class Solution {
		public String largestPalindrome(int n, int k) {
			long[] a = new long[100005];
			a[0] = 1;
			for(int i = 1;i < 100005;i++){
				a[i] = a[i-1] * 10 % k;
			}
			boolean[][] dp = new boolean[k][(n-1)/2+1];
			dp[0][0] = true;
			for(int i = 0;i < (n-1)/2;i++){
				int t = (int)(i == 0 && n/2*2 != n ? a[n/2] : ((a[(n-1)/2-i] + a[n/2+i]) % k));
				for(int u = 0;u < k;u++) {
					if(dp[u][i]){
						for(int j = 0;j <= 9;j++) {
							dp[(u + t * j) % k][i + 1] = true;
						}
					}
				}
			}
			int kk = 0;
			char[] ret = new char[n];
			for(int i = (n-1)/2;i >= 0;i--){
				int t = (int)(i == 0 && n/2*2 != n ? a[n/2] : ((a[(n-1)/2-i] + a[n/2+i]) % k));
				for(int j = 9;j >= 0;j--) {
					int u = (kk-t*j) % k;
					if(u < 0)u += k;
					if (dp[u][i]){
						ret[n/2+i] = ret[(n-1)/2-i] = (char)('0'+j);
						kk = u;
						break;
					}
				}
			}
			return new String(ret);
		}
	}